cmd/local-gadget/profile/cpu.go

```diff
@@ -0,0 +1,88 @@
+// Copyright 2022 The Inspektor Gadget authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package profile
+
+import (
+	"github.com/spf13/cobra"
+
+	commonprofile "github.com/inspektor-gadget/inspektor-gadget/cmd/common/profile"
+	commonutils "github.com/inspektor-gadget/inspektor-gadget/cmd/common/utils"
+	"github.com/inspektor-gadget/inspektor-gadget/cmd/local-gadget/utils"
+	containercollection "github.com/inspektor-gadget/inspektor-gadget/pkg/container-collection"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/gadget-collection/gadgets/profile"
+	cpuTracer "github.com/inspektor-gadget/inspektor-gadget/pkg/gadgets/profile/cpu/tracer"
+	cpuTypes "github.com/inspektor-gadget/inspektor-gadget/pkg/gadgets/profile/cpu/types"
+	localgadgetmanager "github.com/inspektor-gadget/inspektor-gadget/pkg/local-gadget-manager"
+)
+
+func newCPUCmd() *cobra.Command {
+	var commonFlags utils.CommonFlags
+	var cpuFlags commonprofile.CPUFlags
+
+	runCmd := func(*cobra.Command, []string) error {
+		if cpuFlags.ProfileUserOnly && cpuFlags.ProfileKernelOnly {
+			return commonutils.WrapInErrArgsNotSupported("-U and -K can't be used at the same time")
+		}
+
+		parser, err := commonutils.NewGadgetParserWithRuntimeInfo(&commonFlags.OutputConfig, cpuTypes.GetColumns())
+		if err != nil {
+			return commonutils.WrapInErrParserCreate(err)
+		}
+
+		cpuGadget := &ProfileGadget{
+			commonFlags: &commonFlags,
+			parser: &commonprofile.CPUParser{
+				GadgetParser: *parser,
+				OutputConfig: commonFlags.OutputConfig,
+				CPUFlags:     &cpuFlags,
+			},
+			inProgressMsg: "Capturing stack traces",
+			createAndRunTracer: func() (profile.Tracer, error) {
+				localGadgetManager, err := localgadgetmanager.NewManager(commonFlags.RuntimeConfigs)
+				if err != nil {
+					return nil, commonutils.WrapInErrManagerInit(err)
+				}
+				defer localGadgetManager.Close()
+
+				// TODO: Improve filtering, see further details in
+				// https://github.com/inspektor-gadget/inspektor-gadget/issues/644.
+				containerSelector := containercollection.ContainerSelector{
+					Name: commonFlags.Containername,
+				}
+
+				// Create mount namespace map to filter by containers
+				mountnsmap, err := localGadgetManager.CreateMountNsMap(containerSelector)
+				if err != nil {
+					return nil, commonutils.WrapInErrManagerCreateMountNsMap(err)
+				}
+				defer localGadgetManager.RemoveMountNsMap()
```

Comment:
Mmmmm I think this should go within the `runCmd` context to avoid `defers` are called after `NewTracer` returns.

Reply:
Now that you have mentioned it: Doesn't it have to stay inside of the `Run()` method of `cmd/local-gadget/profile/profile.go`?

The `mountnsmap` and `localGadgetMangaer` needs to exist while the tracer is running, or not?

Reply:
Yep, that's what I am saying that we need to move it one level up, in the `runCmd` instead of `createAndRunTracer` context. In this way, they will exist while the gadget is running.

Reply:
Correct, didn't see that the first time. `cpuGadget.Run()` calls the `Run()` of the `ProfileGadget` so everything is good